File: src/tools/convert_sidre_protocol.cpp
// Copyright (c) 2017-2021, Lawrence Livermore National Security, LLC and
// other Axom Project Developers. See the top-level LICENSE file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)

/**
 * \file convert_sidre_protocol.cpp
 * \brief This file contains a utility to convert a Sidre datastore
 *        from the sidre_hdf5 protocol to another supported protocol.
 *
 * Users must supply a path to a sidre_hdf5 rootfile and base name for
 * the output datastores.  Optional command line arguments include
 * a '--protocol' option (the default is 'json')
 * and a '--strip' option to truncate the array data to at most N elements.
 * The strip option also prepends each array with its original size, the new
 * size and a filler entry of 0 for integer arrays or nan for floating point
 * arrays. E.g. if the array had 6 entries [1.01. 2.02, 3.03, 4.04, 5.05, 6.06]
 * and the user passed in --strip 3, the array would be converted to
 * [6, 3, nan, 1.01, 2.02, 3.03].
 *
 * \note The strip option is intended as a temporary solution to truncating
 * a dataset to allow easier debugging.  In the future, we intend to separate
 * the conversion and truncation/display functionality into separate utilities.
 */

#include "axom/config.hpp"
#include "fmt/fmt.hpp"
#include "axom/slic.hpp"
#include "axom/sidre.hpp"
#include "axom/slam.hpp"

#include "mpi.h"
#include "CLI11/CLI11.hpp"

#include <limits>   // for numeric_limits<int>
#include <cstdlib>  // for atoi
#include <sstream>  // for stringstream

namespace sidre = axom::sidre;
namespace slam = axom::slam;
namespace slic = axom::slic;

using PosType = slam::DefaultPositionType;
using ElemType = sidre::IndexType;

using SzPol = slam::policies::RuntimeSize<PosType>;
using OffPol = slam::policies::ZeroOffset<PosType>;
using StrPol = slam::policies::RuntimeStride<PosType>;
using ViewSet = slam::OrderedSet<PosType, ElemType, SzPol, OffPol, StrPol>;

void setupLogging();
void teardownLogging();

/** Simple structure to hold the parsed command line arguments */
struct CommandLineArguments
{
  static const std::set<std::string> s_validProtocols;

  std::string m_inputName;
  std::string m_outputName;
  std::string m_protocol;
  int m_numStripElts;

  CommandLineArguments()
    : m_inputName("")
    , m_outputName("")
    , m_protocol("json")
    , m_numStripElts(-1)
  { }

  void parse(int argc, char** argv, CLI::App& app);

  bool shouldStripData() const { return m_numStripElts >= 0; }

  /**  Returns the maximum allowed elements in a view of the output datastore */
  int maxEltsPerView() const
  {
    return shouldStripData() ? m_numStripElts : std::numeric_limits<int>::max();
  }
};

const std::set<std::string> CommandLineArguments::s_validProtocols(
  {"json",
   "sidre_hdf5",
   "sidre_conduit_json",
   "sidre_json",
   "conduit_hdf5",
   "conduit_bin",
   "conduit_json"});

/** Terminates execution */
void quitProgram(int exitCode = 0)
{
  teardownLogging();
  MPI_Finalize();
  exit(exitCode);
}

/** Parse the command line arguments */
void CommandLineArguments::parse(int argc, char** argv, CLI::App& app)
{
  app
    .add_option("-i,--input",
                m_inputName,
                "Filename of input sidre-hdf5 datastore")
    ->required()
    ->check(CLI::ExistingFile);

  app
    .add_option("-o,--output",
                m_outputName,
                "Filename of output datastore (without extension)")
    ->required();

  app
    .add_option("-p,--protocol",
                m_protocol,
                "Desired protocol for output datastore")
    ->capture_default_str()
    ->check(CLI::IsMember {CommandLineArguments::s_validProtocols});

  app
    .add_option(
      "-s,--strip",
      m_numStripElts,
      "If provided, output arrays will be stripped to first N entries")
    ->check(CLI::PositiveNumber);

  bool verboseOutput = false;
  app.add_flag("-v,--verbose", verboseOutput, "Sets output to verbose")
    ->capture_default_str();

  app.get_formatter()->column_width(35);

  // Could throw an exception
  app.parse(argc, argv);

  if(verboseOutput)
  {
    slic::setLoggingMsgLevel(slic::message::Debug);
  }
}

/**
 * \brief Allocate storage for external data of the input datastore
 *
 * Iterates recursively through the views and groups of the provided group to
 * find the external data views and allocates the required storage within the
 * extPtrs vector
 *
 * \param grp  The group to traverse
 * \param extPtrs [out] A vector to hold pointers to the allocated data
 *
 * \note Also initializes the data in each allocated array to zeros
 */
void allocateExternalData(sidre::Group* grp, std::vector<void*>& extPtrs)
{
  // for each view
  for(auto idx = grp->getFirstValidViewIndex(); sidre::indexIsValid(idx);
      idx = grp->getNextValidViewIndex(idx))
  {
    sidre::View* view = grp->getView(idx);
    if(view->isExternal())
    {
      SLIC_DEBUG("External view " << view->getPathName() << " has "
                                  << view->getNumElements() << " elements "
                                  << "(" << view->getTotalBytes() << " bytes).");

      const int idx = extPtrs.size();
      const int sz = view->getTotalBytes();
      extPtrs.push_back(new char[sz]);
      std::memset(extPtrs[idx], 0, sz);
      view->setExternalDataPtr(extPtrs[idx]);
    }
  }

  // for each group
  for(auto idx = grp->getFirstValidGroupIndex(); sidre::indexIsValid(idx);
      idx = grp->getNextValidGroupIndex(idx))
  {
    allocateExternalData(grp->getGroup(idx), extPtrs);
  }
}

/**
 * \brief Shifts the data to the right by three elements
 *
 * The new first value will be the size of the original array.
 * The next value will be the number of retained elements and the
 * third value will be 0 for integer data and Nan for float data.
 * This is followed by the values in the truncated original dataset.
 *
 * \note This function creates a copy of the data since there could be
 * several views in the original dataset pointing to the same memory
 *
 * \param view The array view on which we are operating
 * \param origSize The size of the original array
 */
template <typename sidre_type>
void modifyFinalValuesImpl(sidre::View* view, int origSize)
{
  sidre_type* arr = view->getData();
  int sz = view->getNumElements();

  // Uses a Slam set to help manage the indirection to the view data
  // Note: offset is zero since getData() already accounts for the offset
  ViewSet viewInds = ViewSet::SetBuilder().size(sz).stride(view->getStride());

#ifdef AXOM_DEBUG
  {
    fmt::memory_buffer out;
    for(auto i : viewInds.positions())
    {
      fmt::format_to(out,
                     "\n\ti: {0}; index: {1}; arr[{1}] = {2}",
                     i,
                     viewInds[i],
                     arr[viewInds[i]]);
    }
    SLIC_DEBUG("Before truncation" << fmt::to_string(out));
  }
#endif

  // Create a new buffer for copied data
  auto* ds = view->getOwningGroup()->getDataStore();
  auto type = view->getTypeID();
  auto newSz = sz + 3;
  auto* buff = ds->createBuffer(type, newSz)->allocate();

  // Explicitly set the first two elements and copy elements over
  sidre_type* newArr = buff->getData();

  newArr[0] = static_cast<sidre_type>(origSize);
  newArr[1] = static_cast<sidre_type>(sz);
  newArr[2] = std::numeric_limits<sidre_type>::quiet_NaN();
  for(auto i : viewInds.positions())
  {
    newArr[i + 3] = arr[viewInds[i]];
  }

  // Update view's buffer to the new data
  view->detachBuffer();
  view->attachBuffer(type, newSz, buff);

#ifdef AXOM_DEBUG
  {
    fmt::memory_buffer out;
    for(auto i : ViewSet(newSz).positions())
    {
      fmt::format_to(out, "\n\ti: {0}; arr[{0}] = {1}", i, newArr[i]);
    }
    SLIC_DEBUG("After truncation" << fmt::to_string(out));
  }
#endif
}

void modifyFinalValues(sidre::View* view, int origSize)
{
  SLIC_DEBUG("Truncating view " << view->getPathName());

  switch(view->getTypeID())
  {
  case sidre::INT8_ID:
    modifyFinalValuesImpl<axom::int8>(view, origSize);
    break;
  case sidre::INT16_ID:
    modifyFinalValuesImpl<axom::int16>(view, origSize);
    break;
  case sidre::INT32_ID:
    modifyFinalValuesImpl<axom::int32>(view, origSize);
    break;
  case sidre::INT64_ID:
    modifyFinalValuesImpl<axom::int64>(view, origSize);
    break;
  case sidre::UINT8_ID:
    modifyFinalValuesImpl<axom::uint8>(view, origSize);
    break;
  case sidre::UINT16_ID:
    modifyFinalValuesImpl<axom::uint16>(view, origSize);
    break;
  case sidre::UINT32_ID:
    modifyFinalValuesImpl<axom::uint32>(view, origSize);
    break;
  case sidre::UINT64_ID:
    modifyFinalValuesImpl<axom::uint64>(view, origSize);
    break;
  case sidre::FLOAT32_ID:
    modifyFinalValuesImpl<axom::float32>(view, origSize);
    break;
  case sidre::FLOAT64_ID:
    modifyFinalValuesImpl<axom::float64>(view, origSize);
    break;
  default:
    break;
  }
}

/**
 * \brief Recursively traverse views and groups in grp and truncate views to
 * have at most maxSize+2 elements.
 *
 * Within the truncated arrays, the first element will be the size of the
 * original array and the second will
 * be 0 for integers and nan for floating points.
 * This will be followed by (at most) the first maxSize elements of the original
 * array
 */
void truncateBulkData(sidre::Group* grp, int maxSize)
{
  // Add two to maxSize
  for(auto idx = grp->getFirstValidViewIndex(); sidre::indexIsValid(idx);
      idx = grp->getNextValidViewIndex(idx))
  {
    sidre::View* view = grp->getView(idx);
    bool isArray = view->hasBuffer() || view->isExternal();

    if(isArray)
    {
      const int numOrigElts = view->getNumElements();
      const int newSize = std::min(maxSize, numOrigElts);

      if(view->hasBuffer() && numOrigElts > newSize)
      {
        const int viewStride = view->getStride();
        const int viewOffset = view->getOffset();
        view->apply(newSize, viewOffset, viewStride);
      }
      // external
      else if(view->isExternal() && numOrigElts > newSize)
      {
        void* dataPtr = view->getVoidPtr();
        view->setExternalDataPtr(view->getTypeID(), newSize, dataPtr);
      }

      modifyFinalValues(view, numOrigElts);
    }
  }

  // for each group
  for(auto idx = grp->getFirstValidGroupIndex(); sidre::indexIsValid(idx);
      idx = grp->getNextValidGroupIndex(idx))
  {
    truncateBulkData(grp->getGroup(idx), maxSize);
  }
}

/** Sets up the logging using lumberjack */
void setupLogging()
{
  slic::initialize();
  slic::setLoggingMsgLevel(slic::message::Info);

#ifdef AXOM_USE_LUMBERJACK
  std::string rankStr = "[<RANK>]";
#else
  std::string rankStr = "";
#endif

  // Formatting for warning, errors and fatal message
  std::stringstream wefFmt;
  wefFmt << "\n***********************************\n"
         << rankStr << "[<LEVEL> in line <LINE> of file <FILE>]\n"
         << "MESSAGE=<MESSAGE>\n"
         << "***********************************\n";
  std::string wefFormatStr = wefFmt.str();

  // Simple formatting for debug and info messages
  std::string diFormatStr = rankStr + "[<LEVEL>]: <MESSAGE>\n";

  slic::LogStream* wefStream;
  slic::LogStream* diStream;

#ifdef AXOM_USE_LUMBERJACK
  const int ranksLimit = 16;
  wefStream = new slic::LumberjackStream(&std::cout,
                                         MPI_COMM_WORLD,
                                         ranksLimit,
                                         wefFormatStr);
  diStream =
    new slic::LumberjackStream(&std::cout, MPI_COMM_WORLD, ranksLimit, diFormatStr);
#else
  wefStream = new slic::GenericOutputStream(&std::cout, wefFormatStr);
  diStream = new slic::GenericOutputStream(&std::cout, diFormatStr);
#endif

  slic::addStreamToMsgLevel(wefStream, slic::message::Error);
  slic::addStreamToMsgLevel(wefStream, slic::message::Warning);
  slic::addStreamToMsgLevel(diStream, slic::message::Info);
  slic::addStreamToMsgLevel(diStream, slic::message::Debug);

  // the following is helpful for debugging
  // slic::debug::checksAreErrors = true;
}

/** Finalizes logging and flushes streams */
void teardownLogging() { slic::finalize(); }

int main(int argc, char* argv[])
{
  MPI_Init(&argc, &argv);

  int my_rank;
  MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);

  setupLogging();

  // parse the command line arguments
  CommandLineArguments args;
  CLI::App app {"Sidre protocol converter"};

  try
  {
    args.parse(argc, argv, app);
  }
  catch(const CLI::ParseError& e)
  {
    int retval = -1;
    if(my_rank == 0)
    {
      retval = app.exit(e);
    }
    MPI_Bcast(&retval, 1, MPI_INT, 0, MPI_COMM_WORLD);
    quitProgram(retval);
  }

  // Load the original datastore
  SLIC_INFO("Loading datastore from " << args.m_inputName);
  sidre::DataStore ds;
  sidre::IOManager manager(MPI_COMM_WORLD);
  manager.read(ds.getRoot(), args.m_inputName);
  int num_files = manager.getNumFilesFromRoot(args.m_inputName);

  // Restore any external data pointers
  SLIC_INFO("Loading external data from datastore");
  std::vector<void*> externalDataPointers;
  allocateExternalData(ds.getRoot(), externalDataPointers);
  manager.loadExternalData(ds.getRoot(), args.m_inputName);

  slic::flushStreams();

  // Internal processing
  if(args.shouldStripData())
  {
    const int numElts = args.maxEltsPerView();
    SLIC_INFO("Truncating views to at most " << numElts << " elements.");

    truncateBulkData(ds.getRoot(), numElts);

    // Add a string view to the datastore to indicate that we modified the data
    std::stringstream sstr;
    sstr << "This datastore was created by axom's 'convert_sidre_protocol' "
         << "utility with option '--strip " << numElts << "'. "
         << "To simplify debugging, the bulk data in this datastore has been "
         << "truncated to have at most " << numElts << " original values "
         << "per array. Three values have been prepended to each array: "
         << "the size of the original array, the number of retained elements "
         << "and a zero/Nan.";

    ds.getRoot()->createViewString("Note", sstr.str());
  }

  // Write out datastore to the output file in the specified protocol
  SLIC_INFO("Writing out datastore in '"
            << args.m_protocol << "' protocol to file(s) with base name "
            << args.m_outputName);
  manager.write(ds.getRoot(), num_files, args.m_outputName, args.m_protocol);

  // Free up memory associated with external data
  for(std::vector<void*>::iterator it = externalDataPointers.begin();
      it != externalDataPointers.end();
      ++it)
  {
    delete[] static_cast<char*>(*it);
    *it = nullptr;
  }

  teardownLogging();
  MPI_Finalize();
  return 0;
}

File: src/tools/convert_sidre_protocol.py
# Copyright (c) Lawrence Livermore National Security, LLC and other
# Axom Project Contributors. See top-level LICENSE and COPYRIGHT
# files for dates and other details.
#
# SPDX-License-Identifier: (BSD-3-Clause)
"""
Convert a Sidre datastore or Conduit Blueprint mesh to another protocol.

Users must supply a path to an input rootfile and base name for the output.
Sidre datastore conversion remains the default. Use ``--input-type blueprint``
to convert Conduit Blueprint meshes, including HDF5 Blueprint root files.
Optional command line arguments include a ``--protocol`` option (the default is ``json``)
and a ``--strip`` option to truncate the array data to at most N elements.
The strip option also prepends each array with its original size, the new
size and a filler entry of 0 for integer arrays or nan for floating point
arrays. E.g. if the array had 6 entries [1.01, 2.02, 3.03, 4.04, 5.05, 6.06]
and the user passed in ``--strip 3``, the array would be converted to
[6, 3, nan, 1.01, 2.02, 3.03].

For Blueprint meshes, ``--strip`` applies to numeric array leaves with more than
one element and leaves scalar/string metadata intact.
A ``state/Note`` node is added to each output domain to record that the mesh was stripped.

The resulting stripped output is intended for debugging, not for use as a valid mesh.
In the future, the conversion and truncation/display functionality may be
separated into distinct utilities.
"""

from __future__ import annotations

import sys
import argparse
import math
from pathlib import Path

import numpy as np
import axom.sidre as sidre

SIDRE_PROTOCOLS = (
    "json",
    "sidre_hdf5",
    "sidre_conduit_json",
    "sidre_json",
    "conduit_hdf5",
    "conduit_bin",
    "conduit_json",
)

BLUEPRINT_PROTOCOLS = (
    "hdf5",
    "json",
    "yaml",
    "conduit_bin",
    "conduit_hdf5",
    "conduit_json",
)

VALID_PROTOCOLS = tuple(dict.fromkeys(SIDRE_PROTOCOLS + BLUEPRINT_PROTOCOLS))


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Sidre/Blueprint protocol converter")
    parser.add_argument(
        "-i",
        "--input",
        required=True,
        help="Filename of input sidre-hdf5 datastore or Blueprint mesh root file",
    )
    parser.add_argument(
        "-o",
        "--output",
        required=True,
        help="Filename of output datastore/mesh (without extension)",
    )
    parser.add_argument(
        "--input-type",
        choices=("sidre", "blueprint"),
        default="sidre",
        help="Type of input file to convert",
    )
    parser.add_argument(
        "-p",
        "--protocol",
        default="json",
        choices=VALID_PROTOCOLS,
        help="Desired output protocol; valid protocols depend on --input-type",
    )
    parser.add_argument(
        "-s",
        "--strip",
        type=int,
        default=None,
        help="If provided, output arrays will be stripped to first N entries",
    )
    parser.add_argument(
        "-v",
        "--verbose",
        action="store_true",
        help="Sets output to verbose",
    )
    return parser.parse_args()


def initialize_mpi() -> tuple[object | None, bool, int, int]:
    if not sidre.AXOM_ENABLE_MPI:
        return None, False, 1, 0

    try:
        from mpi4py import MPI
    except ImportError as exc:
        raise RuntimeError(
            "convert_sidre_protocol.py requires mpi4py when Axom is built with MPI support",
        ) from exc

    initialized_mpi = False
    if not MPI.Is_initialized():
        MPI.Init()
        initialized_mpi = True

    return MPI, initialized_mpi, MPI.COMM_WORLD.Get_size(), MPI.COMM_WORLD.Get_rank()


#
# Allocate storage for external data of the input datastore.
#
# Iterates recursively through the views and groups of the provided group to
# find the external data views and allocates the required storage within the
# holders list.
#
# Also initializes the data in each allocated array to zeros.
#
def allocate_external_data(group: sidre.Group, holders: list[np.ndarray], verbose: bool) -> None:
    # for each view
    for view in group.views():
        if view.isExternal():
            if verbose:
                print(
                    f"Allocating external storage for {view.getPathName()} "
                    f"({view.getNumElements()} elements, {view.getTotalBytes()} bytes)", )
            storage = np.zeros(view.getTotalBytes(), dtype=np.uint8)
            view.setExternalData(storage)
            holders.append(storage)

    # for each group
    for child in group.groups():
        allocate_external_data(child, holders, verbose)


#
# Shift the data to the right by three elements.
#
# The new first value will be the size of the original array.
# The next value will be the number of retained elements and the
# third value will be 0 for integer data and Nan for float data.
# This is followed by the values in the truncated original dataset.
#
# This function creates a copy of the data since there could be
# several views in the original dataset pointing to the same memory.
#
def modify_final_values(
    view: sidre.View,
    original_size: int,
    retained_size: int | None = None,
) -> None:
    flattened = np.asarray(view.getDataArray()).reshape(-1)
    if retained_size is None:
        retained = flattened
    else:
        retained = flattened[:retained_size]

    retained_size = int(retained.size)

    # Create a new buffer for copied data.
    datastore = view.getOwningGroup().getDataStore()
    type_id = view.getTypeID()
    new_size = retained_size + 3
    buffer = datastore.createBuffer(type_id, new_size)
    buffer.allocate()

    # Explicitly set the first two elements and copy elements over.
    new_array = np.asarray(buffer.getDataArray()).reshape(-1)
    np.copyto(new_array[:2], np.asarray([original_size, retained_size]), casting="unsafe")
    new_array[2] = math.nan if np.issubdtype(new_array.dtype, np.floating) else 0
    if retained_size > 0:
        np.copyto(new_array[3:], retained, casting="unsafe")

    # Update view's buffer to the new data.
    # The C++ utility uses detachBuffer() here because this path is valid for
    # buffer-backed views. In Python we also need to support external views, so
    # we make the state transition explicit before attaching the new buffer.
    if view.hasBuffer():
        view.attachBuffer(None)
    elif view.isExternal():
        view.setExternalData(None)

    view.attachBuffer(type_id, new_size, buffer)


#
# Recursively traverse views and groups in group and truncate views to
# have at most max_size elements.
#
# Within the truncated arrays, the first element will be the size of the
# original array, the second will be the number of retained elements and the
# third will be 0 for integers or nan for floating points.
# This will be followed by at most the first max_size elements of the
# original array.
#
def truncate_bulk_data(group: sidre.Group, max_size: int, verbose: bool) -> None:
    # for each view
    for view in group.views():
        is_array = view.hasBuffer() or view.isExternal()

        if is_array:
            original_size = view.getNumElements()
            retained_size = min(max_size, original_size)

            if view.hasBuffer() and original_size > retained_size:
                view.apply(retained_size, view.getOffset(), view.getStride())
            elif view.isExternal() and original_size > retained_size:
                data = np.asarray(view.getDataArray()).reshape(-1)
                view.setExternalData(view.getTypeID(), retained_size, data)

            if verbose:
                print(
                    f"Truncating view {view.getPathName()} from {original_size} to {retained_size}",
                )
            modify_final_values(view, original_size, retained_size)

    # for each group
    for child in group.groups():
        truncate_bulk_data(child, max_size, verbose)


def blueprint_protocol(protocol: str) -> str:
    if protocol == "conduit_hdf5":
        return "hdf5"
    if protocol in BLUEPRINT_PROTOCOLS:
        return protocol

    valid = ", ".join(BLUEPRINT_PROTOCOLS)
    raise RuntimeError(
        f"Protocol '{protocol}' is not valid for Blueprint mesh output. "
        f"Use one of: {valid}", )


def sidre_protocol(protocol: str) -> str:
    if protocol in SIDRE_PROTOCOLS:
        return protocol

    valid = ", ".join(SIDRE_PROTOCOLS)
    raise RuntimeError(
        f"Protocol '{protocol}' is not valid for Sidre datastore output. "
        f"Use one of: {valid}", )


def blueprint_domain_count(mesh) -> int:
    if mesh.has_path("coordsets") and mesh.has_path("topologies"):
        return 1
    return mesh.number_of_children()


def strip_note(data_kind: str, max_size: int) -> str:
    return (f"This {data_kind} was created by axom's 'convert_sidre_protocol' utility "
            f"with option '--strip {max_size}'. To simplify debugging, the bulk "
            f"data in this {data_kind} has been truncated to have at most {max_size} "
            "original values per array. Three values have been prepended to each array: "
            "the size of the original array, the number of retained elements and a zero/Nan.")


def add_blueprint_strip_note(mesh, note: str) -> None:
    if mesh.has_path("coordsets") and mesh.has_path("topologies"):
        mesh["state/Note"] = note
        return

    for child_idx in range(mesh.number_of_children()):
        domain = mesh.child(child_idx)
        if domain.has_path("coordsets") and domain.has_path("topologies"):
            domain["state/Note"] = note


def truncate_conduit_numeric_array(node, max_size: int, verbose: bool) -> int:
    dtype = node.dtype()
    original_size = dtype.number_of_elements()
    if not dtype.is_number() or original_size <= 1:
        return 0

    retained_size = min(max_size, original_size)
    values = np.asarray(node.value()).reshape(-1)
    retained = values[:retained_size].copy()

    new_values = np.empty(retained_size + 3, dtype=values.dtype)
    np.copyto(new_values[:2], np.asarray([original_size, retained_size]), casting="unsafe")
    new_values[2] = math.nan if np.issubdtype(new_values.dtype, np.floating) else 0
    if retained_size > 0:
        np.copyto(new_values[3:], retained, casting="unsafe")

    if verbose:
        print(f"Truncating node {node.path()} from {original_size} to {retained_size}")

    node.reset()
    node.set(new_values)
    return 1


def truncate_conduit_bulk_data(node, max_size: int, verbose: bool) -> int:
    if node.number_of_children() == 0:
        return truncate_conduit_numeric_array(node, max_size, verbose)

    truncated_count = 0
    for child_idx in range(node.number_of_children()):
        truncated_count += truncate_conduit_bulk_data(node.child(child_idx), max_size, verbose)
    return truncated_count


def convert_blueprint_mesh(args: argparse.Namespace, MPI: object | None, comm_size: int,
                           rank: int) -> int:
    import conduit
    import conduit.blueprint
    import conduit.relay.io.blueprint

    input_path = Path(args.input)
    protocol = blueprint_protocol(args.protocol)
    mesh = conduit.Node()

    if MPI is not None and comm_size > 1:
        import conduit.blueprint.mpi.mesh
        import conduit.relay.mpi.io.blueprint

        comm = MPI.COMM_WORLD.py2f()
        if rank == 0:
            print(f"Loading Blueprint mesh from {input_path} on {comm_size} MPI rank(s)", )
        conduit.relay.mpi.io.blueprint.load_mesh(mesh, str(input_path), comm)

        info = conduit.Node()
        valid = conduit.blueprint.mpi.mesh.verify(mesh, info, comm)
        local_domains = blueprint_domain_count(mesh)
        total_domains = MPI.COMM_WORLD.allreduce(local_domains)
        if rank == 0:
            print(f"Input Blueprint mesh layout: {total_domains} domain(s)")
        if not valid:
            raise RuntimeError(f"Input Blueprint mesh failed verification:\n{info.to_yaml()}")

        if args.strip is not None:
            if rank == 0:
                print(f"Truncating numeric Blueprint arrays to at most {args.strip} elements.")
            local_truncated = truncate_conduit_bulk_data(mesh, args.strip, args.verbose)
            total_truncated = MPI.COMM_WORLD.allreduce(local_truncated)
            if rank == 0:
                print(f"Truncated {total_truncated} numeric Blueprint array(s).")
            add_blueprint_strip_note(mesh, strip_note("Blueprint mesh", args.strip))

        if rank == 0:
            print(
                f"Writing out Blueprint mesh in '{protocol}' protocol to file(s) "
                f"with base name {args.output}", )
        conduit.relay.mpi.io.blueprint.save_mesh(mesh, args.output, comm, protocol)
    else:
        print(f"Loading Blueprint mesh from {input_path}")
        conduit.relay.io.blueprint.load_mesh(mesh, str(input_path))

        info = conduit.Node()
        valid = conduit.blueprint.mesh.verify(mesh, info)
        domains = blueprint_domain_count(mesh)
        print(f"Input Blueprint mesh layout: {domains} domain(s)")
        if not valid:
            raise RuntimeError(f"Input Blueprint mesh failed verification:\n{info.to_yaml()}")

        if args.strip is not None:
            print(f"Truncating numeric Blueprint arrays to at most {args.strip} elements.")
            truncated_count = truncate_conduit_bulk_data(mesh, args.strip, args.verbose)
            print(f"Truncated {truncated_count} numeric Blueprint array(s).")
            add_blueprint_strip_note(mesh, strip_note("Blueprint mesh", args.strip))

        print(
            f"Writing out Blueprint mesh in '{protocol}' protocol to file(s) "
            f"with base name {args.output}", )
        conduit.relay.io.blueprint.save_mesh(mesh, args.output, protocol)

    return 0


def convert_sidre_datastore(args: argparse.Namespace, comm_size: int) -> int:
    if not sidre.AXOM_ENABLE_MPI:
        raise RuntimeError("sidre.IOManager bindings require an MPI-enabled Axom build")

    protocol = sidre_protocol(args.protocol)
    input_path = Path(args.input)
    manager = sidre.IOManager()
    datastore = sidre.DataStore()
    root = datastore.getRoot()

    num_files = manager.getNumFilesFromRoot(str(input_path))
    num_groups = manager.getNumGroupsFromRoot(str(input_path))

    print(
        "Input datastore layout: "
        f"{num_groups} rank group(s) across {num_files} file(s); "
        f"running on {comm_size} MPI rank(s)", )

    if comm_size != num_groups:
        print(
            "Warning: current MPI size does not match the input datastore rank count. "
            "sidre_hdf5 supports some rank-count mismatches, but not every layout.", )
        if comm_size < num_groups and num_files != num_groups:
            print(
                "Warning: this run has fewer MPI ranks than the input datastore. "
                "Sidre only supports that case for file-per-rank sidre_hdf5 datasets "
                f"(number_of_files == number_of_trees, but here {num_files} != {num_groups}), "
                "so IOManager.read() is expected to fail.", )
        elif comm_size < num_groups:
            print(
                "Warning: this run has fewer MPI ranks than the input datastore, but the "
                "dataset is file-per-rank "
                f"({num_files} file(s) for {num_groups} rank group(s)), so Sidre can load it. "
                "In that reduced-rank case, one output rank may absorb data from multiple input "
                "ranks, and the loaded hierarchy may be reshaped under "
                "'rank_%07d/sidre_input' groups.", )

    print(f"Loading datastore from {input_path}")
    manager.read(root, str(input_path))

    print("Loading external data from datastore")
    external_holders: list[np.ndarray] = []
    allocate_external_data(root, external_holders, args.verbose)
    manager.loadExternalData(root, str(input_path))

    if args.strip is not None:
        print(f"Truncating views to at most {args.strip} elements.")
        truncate_bulk_data(root, args.strip, args.verbose)
        root.createViewString("Note", strip_note("datastore", args.strip))

    print(
        f"Writing out datastore in '{protocol}' protocol to file(s) with base name {args.output}", )
    manager.write(root, num_files, args.output, protocol)

    return 0


def main() -> int:
    args = parse_args()
    if args.strip is not None and args.strip < 0:
        raise RuntimeError("--strip must be nonnegative")

    MPI, initialized_mpi, comm_size, rank = initialize_mpi()
    try:
        if args.input_type == "blueprint":
            return convert_blueprint_mesh(args, MPI, comm_size, rank)
        return convert_sidre_datastore(args, comm_size)
    finally:
        if MPI is not None and initialized_mpi and not MPI.Is_finalized():
            MPI.Finalize()


if __name__ == "__main__":
    sys.exit(main())
